Incorporate blocking issues feedback on initial e2e environment package

rrennick opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

- Software versions can now be specified with 
```
TRAVIS_MARIADB_VERSION=10.5.3 TRAVIS_PHP_VERSION=7.4.5 WP_VERSION=5.4.1 npm run docker:up
```
- Docker will download and cache a copy of libraries when requesting versions that haven't been used in Docker on a system before. (eg. When WordPress 5.5 is released a new Docker WordPress image will be downloaded.) You will see the downloads logged in the output.
- Ensure Docker is down when making changes to `tests/e2e/config/default.json`. 

### How to test the changes in this Pull Request:

1. Run `npm run build:packages`
2. Try `up` & `down` using combinations of environment variables above including with none set
3. `npm run test:e2e`
4. Copy `tests/e2e/env/config/default.json` to `tests/e2e/config/default.json`
5. Edit `tests/e2e/config/default.json` and change the port in the `url` value
6. `npm run docker:up` and `npm run test:e2e`.
7. Create the following `/tests/e2e/config/jest-puppeteer.config.js`
```
const { useE2EJestPuppeteerConfig } = require( '@woocommerce/e2e-environment' );

const puppeteerConfig = useE2EJestPuppeteerConfig( {
	launch: {
		headless: false,
	}
} );

module.exports = puppeteerConfig;
```
8. `npm run test:e2e`
9. Tests should run with browser (some may fail for timing)
10. Change `headless` to `true`
11. `npm run test:e2e-dev` should run without browser
12. Replace `/tests/e2e/config/jest.config.js` in this branch with the version in `master`
13. `npm run test:e2e`

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

rrennick committed:
add PHP, MariaDB, WP env var version support

rrennick committed:
eliminate project root docker-compose.yaml

rrennick committed:
add port configuration support

rrennick committed:
fix software version checks in Travis

rrennick committed:
update version script to work in both desktop and Travis, add fallback versions

claudiosanches commented on package.json:
```diff
@@ -106,5 +106,8 @@
-  ]
+  ],
+  "dependencies": {
+    "@automattic/puppeteer-utils": "github:Automattic/puppeteer-utils#0f3ec50"
```
I think this intent to be a `devDependencies`

rrennick replied:
@claudiosanches I will double check but IIRC when I was working on this, the package dev dependencies are for development of the package in this repo while the dependencies are the packages it needs to run as a package in a down stream project.

rrennick replied:
@claudiosanches Fixed up in https://github.com/woocommerce/woocommerce/pull/27113/commits/e78beaf5693609140dfcd7ed75ab76e6e631cb3a It was added to the wrong package.json.

rrennick committed:
move jest/puppeteer related packages to environment package

rrennick committed:
use @automattic/puppeteer-utils

rrennick committed:
add local puppeteer config support

rrennick committed:
merge master

rrennick committed:
restore non-dev puppeteer config, move test config to sequencer

rrennick committed:
expose getAppRoot, getTestConfig in package

rrennick committed:
use consistent MySQL configuration values

rrennick committed:
move trash posts/products from sequencer to local config

rrennick committed:
merge master

rrennick committed:
keep browser reset at end of setup

rrennick committed:
update e2e readmes

renatho commented on tests/e2e/env/bin/docker-compose.sh:
```diff
@@ -0,0 +1,44 @@
+	if [[ $WP_VERSION =~ ^[0-9]+\.[0-9]+ ]]; then
+		export WORDPRESS_VERSION=$WP_VERSION
+	else
+		export WORDPRESS_VERSION="5.4.2"
```
WDYT to use `latest` here instead of the fixed version?

rrennick replied:
@renatho If you create a docker container with `wordpress:latest` at version 5.2, docker caches that version as `latest`. When 5.3, .4, etc. are released the container WP version remains 5.2.

renatho replied:
Ah, I got! Thank you for the answer! :)

renatho reviewed (commented):
Thank you for working on this package! We're using it in the Sensei LMS plugin. 😊

rrennick committed:
merge package config with factories changes

rrennick committed:
update package-lock.json

rrennick committed:
remove duplicate readme text

rrennick commented:
@juliaamosova Can you review the instruction changes in the readme files and verify that the instructions are accurate?

juliaamosova reviewed (commented):
@rrennick thanks for working on this! 

I tested this PR and the majority of things worked great for me. Documenting my experience below:

- Building the env running `TRAVIS_MARIADB_VERSION=10.5.3 TRAVIS_PHP_VERSION=7.4.5 WP_VERSION=5.4.1 npm run docker:up` is a bliss. I've tried several combinations and it works really nice. 

- Once I got Docker up and running, I couldn't immediately run the tests. I was getting the following error:

```
 FAIL  ../specs/front-end/front-end-single-product.test.js
  ● Test suite failed to run

    Cannot find module '@woocommerce/model-factories' from 'factories.js'

    Require stack:
      /Users/juliaamosova/src/woocommerce/tests/e2e/utils/factories.js
      /Users/juliaamosova/src/woocommerce/tests/e2e/utils/components.js
      /Users/juliaamosova/src/woocommerce/tests/e2e/specs/front-end/front-end-single-product.test.js

    > 1 | import {
        | ^
      2 | 	AdapterTypes,
      3 | 	initializeUsingBasicAuth,
      4 | 	ModelRegistry,

      at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:299:11)
      at Object.<anonymous> (../utils/factories.js:1:1)
```

I solved it by running  `npm run build:packages`. Documenting it here in case others will encounter the same. 

After that, all the tests run and pass successfully locally for me in the headless mode (by running `npm run test:e2e`). 

- Changing the port in the `url` value by copying and editing `tests/e2e/config/default.json` - works well.

- Running `npm run docker:up` without specifying the environment variables didn't work for me. It tries to pull the latest version of the `mariaDB` which 10.5.5 ([released today on August 10, 2020](https://mariadb.com/kb/en/mariadb-1055-release-notes/)). Getting the following error:

```
julia-mbp:~/src/woocommerce (add/e2e-feedback-blockers * u=)$ npm run docker:up

> woocommerce@4.4.0 docker:up /Users/juliaamosova/src/woocommerce
> npm explore @woocommerce/e2e-environment -- npm run docker:up


> @woocommerce/e2e-environment@0.1.5 docker:up /Users/juliaamosova/src/woocommerce/tests/e2e/env
> ./bin/docker-compose.sh up

WordPress 5.4.2
PHP 7.4.9
MariaDB 10.5.5
Creating network "env_default" with the default driver
Creating volume "env_db" with default driver
Creating volume "env_wordpress" with default driver
Pulling db (mariadb:10.5.5)...
ERROR: manifest for mariadb:10.5.5 not found: manifest unknown: manifest unknown
Docker exit code: 1
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @woocommerce/e2e-environment@0.1.5 docker:up: `./bin/docker-compose.sh up`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @woocommerce/e2e-environment@0.1.5 docker:up script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/juliaamosova/.npm/_logs/2020-08-10T22_38_01_838Z-debug.log
npm ERR! weird error 1
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! woocommerce@4.4.0 docker:up: `npm explore @woocommerce/e2e-environment -- npm run docker:up`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the woocommerce@4.4.0 docker:up script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/juliaamosova/.npm/_logs/2020-08-10T22_38_01_875Z-debug.log
julia-mbp:~/src/woocommerce (add/e2e-feedback-blockers * u=)$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

Noting that `TRAVIS_MARIADB_VERSION=10.5.5 TRAVIS_PHP_VERSION=7.4.5 WP_VERSION=5.4.2 npm run docker:up` produces the same error. Could it be related to the release of `mariaDB` itself?

- Confirm that after creating `/tests/e2e/config/jest-puppeteer.config.js` and setting `headless: false` in it, the tests run with the browser open as they should. 

- Confirm that after setting `headless: true` in `/tests/e2e/config/jest-puppeteer.config.js`, the tests run headless (without browser open) as they should.

- Confirm that tests run after replacing `/tests/e2e/config/jest.config.js` in this branch with the version in `master`.  

>@juliaamosova Can you review the instruction changes in the readme files and verify that the instructions are accurate?

It looks great to me! Although a bit hard to tell without using this package and the documentation first hand in another repository. 

I think we can keep it as-is (unless someone else wants to add something during the review of this PR) and add missing bits (if any) as we start using the package. 

On another note, empty file `tests/e2e/config/tests-e2e.json` can be removed since it used to serve the same purpose as `tests/e2e/config/default.json` now serves.

rrennick commented:
@juliaamosova Thanks for the feedback :)

> I solved it by running `npm run build:packages`. 

I added this to the test instructions

>  It tries to pull the latest version of the mariaDB which 10.5.5 (released today on August 10, 2020). 

Docker creates container images for each supported combination of Linux, PHP, MariaDB, and WP. It takes a couple days after a new software version is released for the images for the new version to be created. The version environment variables are always defined in the Travis environment so this will only intermittently affect local testing. I may be able to find a way to prevent this happening for the next release.

> On another note, empty file tests/e2e/config/tests-e2e.json can be removed since it used to serve the same purpose as tests/e2e/config/default.json now serves.

I'll update the PR for this.

rrennick committed:
remove unused local test-e2e.json

juliaamosova commented:
>I added this to the test instructions

Great, thanks @rrennick. 

>Docker creates container images for each supported combination of Linux, PHP, MariaDB, and WP. It takes a couple days after a new software version is released for the images for the new version to be created. The version environment variables are always defined in the Travis environment so this will only intermittently affect local testing. I may be able to find a way to prevent this happening for the next release.

That makes sense, thanks for noting it. We can also make a note of it in the docs and suggest to use environment variables in cases like this as an option. 

>I'll update the PR for this.

Thanks!

juliaamosova reviewed (approved):
>I'll update the PR for this.

The changes look good to me, thanks @rrennick! It looks like there is a conflict with `package-lock.json` currently but should be good to merge as soon as the conflict is resolved.

rrennick committed:
update package-lock.json

ObliviousHarmony committed:
Revised the default environment variables to pull the version from the latest docker image

ObliviousHarmony committed:
Merge remote-tracking branch 'origin/master' into add/e2e-feedback-blockers

ObliviousHarmony committed:
Resolved package conflicts

ObliviousHarmony committed:
Added package building to npm install

ObliviousHarmony committed:
Updated package-lock.json files

ObliviousHarmony committed:
Updated lock files & fixed failing E2E test

ObliviousHarmony committed:
Fixed another E2E test

ObliviousHarmony reviewed (commented):
Great work on this @rrennick. I encountered a few problems in my testing but I've gone ahead and fixed all of them. While the Travis build has some kind of error I'll note everything works great locally and all of the tests pass.

- Once the `package-lock.json` file was updated in 17ee223 the tests all started failing. I think we need to start using `nvm` so that we can avoid oddities caused by folks using different NPM versions.
- When testing I immediately encountered the same image related problem as @juliaamosova and so I've addressed that in 16cf8c8 by switching the default to the latest image using `docker image ls <repository>`.
- I've added a `lerna bootstrap` to our `install` event which installs all of the packages as well as builds them. This should avoid ambiguous failures and make it a bit easier to work with our growing number of packages.

I am getting a strange message when running `npm install` but it doesn't seem to actually break anything:
```bash
> puppeteer@2.1.1 install /Users/obliviousharmony/Repositories/WooCommerce/Core/node_modules/@typescript-eslint/typescript-estree/node_modules/puppeteer
> node install.js

internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module 'mime'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/Users/obliviousharmony/Repositories/WooCommerce/Core/node_modules/@typescript-eslint/typescript-estree/node_modules/puppeteer/lib/Page.js:19:14)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)

```
My thinking is that this has something to do with the transitive dependencies of the package and I am able to get rid of this failure by adding `puppeteer` back to the root `package.json`. I didn't push a commit for this however as I was hoping you might have some insight into the failure as well.

As an aside, I'm wondering how you would feel about using `"peerDependencies"` for Jest & Puppeter? Given that the environment is providing these dependencies just to add extensibility to them I think it might make more sense for them to be declared in consuming projects.

tammullen commented:
Thanks @rrennick for all the great work on this

Everything in the test instructions worked for me (thanks @ObliviousHarmony for the extra fixes).

Was so nice to be able to specify the versions, I tried it with various versions including the latest WP `5.5.1`

> Docker will download and cache a copy of libraries when requesting versions that haven't been used in Docker on a system before.

Would we would want to also support other docker tags like `latest`? I tried using `latest` as the WordPress version and although nothing broke it just appeared to use the most recent version I already had pulled (`5.4`)

> As an aside, I'm wondering how you would feel about using "peerDependencies" for Jest & Puppeter? Given that the environment is providing these dependencies just to add extensibility to them I think it might make more sense for them to be declared in consuming projects.

An issue we ran into when using the older WooCommerce e2e framework (the Selenium WebDriver based one) from a consuming project was when the ChromeDriver version started needing to be updated more frequently (ChromeDriver started tracking Chrome versions and you had to use the correct version). 
The older version specified in the core framework would sometimes overwrite the version in the consuming project and there would be a delay/block/overhead while we got the version in core updated etc.
Just thinking ahead if we ever found that frequent updates were needed for Jest/Puppeteer.

rrennick commented:
> I tried using latest as the WordPress version and although nothing broke it just appeared to use the most recent version I already had pulled (5.4)

@tammullen `latest` is the default. If you do not specify a version then the build script automatically looks up what is the latest version of WP, MariaDB & PHP and uses those.

rrennick commented:
> My thinking is that this has something to do with the transitive dependencies of the package and I am able to get rid of this failure by adding puppeteer back to the root package.json.

Both `master` and this branch install two versions of `puppeteer`. I think the `2.1.1` version has the `mime` dependency while the `2.0.0` version doesn't. Other packages are relying on 2.1.1 to provide `mime` which is removed when `2.0.0` is installed.

In https://github.com/woocommerce/woocommerce/pull/27412 I addressed the multiple puppeteer install issue. Today I merged this branch into that PR and updated the set up instructions to include `npm install puppeteer@2.1.1 --save` to prevent Chromium being downloaded into the node_modules folder of each package that uses puppeteer.

If #27412 tests clean we could merge this one then it so master remains in a working state.

rrennick committed:
fix up package.json for consistent puppeteer versions

rrennick committed:
merge master

rrennick committed:
update webpack to latest

rrennick committed:
run grunt directly in travis e2e

ObliviousHarmony committed:
Updated all of the package-lock files using Node v10

ObliviousHarmony committed:
Fixed a failing E2E test

ObliviousHarmony committed:
Merge branch 'master' into add/e2e-feedback-blockers

ObliviousHarmony reviewed (approved):
I've resolved the problems with CI by regenerating all of the `package-lock.json` files using v10. Now that the tests are passing I'll approve the PR.

I'm going to submit a PR to add NVM to our repository and require its usage to avoid issues like this happening again in the future.

rrennick committed:
update travis setup for e2e

rrennick committed:
use better test for creating ready page

rrennick committed:
restore looking up tags from git, add npm install to travis e2e script
